dawn/src/dawn/Optimizer/PassFixVersionedInputFields.cpp

```diff
@@ -0,0 +1,172 @@
+
+  return assignmentStmt;
+}
+
+/// @brief Creates the do-method
+static std::unique_ptr<iir::DoMethod>
+createDoMethod(int assignmentID, int assigneeID,
+               std::shared_ptr<iir::StencilInstantiation> const& si,
+               const iir::Interval& interval) {
+  iir::StencilMetaInformation& metadata = si->getMetaData();
+
+  // Create the assignment statement with the new and old variables
+  auto assignmentStmt = createAssignmentStatement(assignmentID, assigneeID, metadata);
+
+  iir::Accesses accesses;
+  accesses.addWriteExtent(assignmentID, iir::Extents{});
+  accesses.addReadExtent(assigneeID, iir::Extents{});
+  assignmentStmt->getData<iir::IIRStmtData>().CallerAccesses = std::move(accesses);
+
+  auto domethod = std::make_unique<iir::DoMethod>(interval, metadata);
+  // TODO: This may not be needed...
+  domethod->setID(si->nextUID());
+
+  domethod->getAST().push_back(std::move(assignmentStmt));
+  domethod->update(iir::NodeUpdateType::level);
+
+  return domethod;
+}
+
+/// @brief Creates the stage in which assignment happens (where the temporary gets filled)
+static std::unique_ptr<iir::Stage>
+createAssignmentStage(int assignmentID, int assigneeID,
+                      std::shared_ptr<iir::StencilInstantiation> const& si,
+                      const iir::Interval& interval) {
+  // TODO: Use the IIRBuilder instead of doing this manually...
+
+  // Create the do-method for the stage
+  auto domethod = createDoMethod(assignmentID, assigneeID, si, interval);
+
+  // Create the stage and add the do method
+  iir::StencilMetaInformation& metadata = si->getMetaData();
+  auto assignmentStage = std::make_unique<iir::Stage>(metadata, si->nextUID());
+  assignmentStage->setExtents(iir::Extents{});
+
+  assignmentStage->addDoMethod(std::move(domethod));
+  assignmentStage->update(iir::NodeUpdateType::level);
+
+  return assignmentStage;
+}
+
+/// @brief Creates a multistage in which assignment happens (where the versioned field gets filled)
+static std::unique_ptr<iir::MultiStage>
+createAssignmentMultiStage(int assignmentID, std::shared_ptr<iir::StencilInstantiation> const& si,
+                           const iir::Interval& interval) {
+  iir::StencilMetaInformation& metadata = si->getMetaData();
+  int assigneeID = metadata.getOriginalVersionOfAccessID(assignmentID);
+
+  auto assignmentStage = createAssignmentStage(assignmentID, assigneeID, si, interval);
+
+  // Create the multistage and add the assignment stage
+  auto ms = std::make_unique<iir::MultiStage>(metadata, iir::LoopOrderKind::Parallel);
+  ms->setID(si->nextUID());
+
+  ms->insertChild(std::move(assignmentStage));
+  ms->update(iir::NodeUpdateType::levelAndTreeAbove);
+
+  return ms;
+}
+
+/// @brief Collects AccessIDs from versioned fields in the IIR.
+struct CollectVersionedIDs : public iir::ASTVisitorForwarding {
+  const iir::StencilMetaInformation& metadata_;
+  std::set<int> versionedAccessIDs;
+
```

Comment:
why not have this set private with a getter?

Reply:
I suppose, but that adds code bulk, and the example I based this on in `PassFieldVersioning.cpp` did not have one... What do you think?